### PR TITLE
[CAMEL-15253] camel-oaipmh - basic implementation

### PR DESCRIPTION
**Pull request by juanksega**
A first draft of the camel-oaipmh component. 

Part of this code was imported from: [https://github.com/cbadenes/camel-oaipmh](camel-oaipmh)

Important changes: 
- Remove oai_dc specific classes.
- Add https configuration.
- Test component on a remote server.
- Basic structure of a Producer.

[ ] Make sure there is a [JIRA issue](https://issues.apache.org/jira/browse/CAMEL) filed for the change (usually before you start working on it).  Trivial changes like typos do not require a JIRA issue.  Your pull request should address just this issue, without pulling in other changes.
[ ] Each commit in the pull request should have a meaningful subject line and body.
[ ] If you're unsure, you can format the pull request title like `[CAMEL-XXX] Fixes bug in camel-file component`, where you replace `CAMEL-XXX` with the appropriate JIRA issue.
[ ] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
[ ] Run `mvn clean install -Psourcecheck` in your module with source check enabled to make sure basic checks pass and there are no checkstyle violations. A more thorough check will be performed on your pull request automatically.
Below are the contribution guidelines:
https://github.com/apache/camel/blob/master/CONTRIBUTING.md

### CODE REVIEW

**Comment by juanksega** [comment]
Hi @DenisIstomin @zregvart @oscerd 

I pulled the last commits from the master branch of apache/camel, also I added a Producer implementation and some tests. Please let me know if you have suggestions.

**Comment by bedlaj** [comment]
Good start. Since this is draft, i will give a few suggestions, instead of full review.

- Please avoid using lombok there are many good reasons to not using that.
- Use jsoup and joda-time versions provided in parent pom.
- In final version, there must be all remote http calls removed, instead setup mock or use testcontainers.
- Thread.sleep in tests can be replaced with proper assertions.
- Never use `assert` keyword in unit tests.
- Do not use `@author` annotations in javadoc.
- Is safe to assume that response will be always UTF-8 and in Zulu timezone?
- Headers should follow CamelOaimph* pattern, insead of oaimph.*
- Endpoint class should contain some field annotated with `@UriPath`
- For consumer use AsyncCallback in AsyncProcessor instead of try-catch-finally (Look eg to WsConsumer for inspiration)
- Is `build-helper-maven-plugin` required? I dont see any custom generated (re)sources.

And some obvious things, which I believe you know about.

- Checkstyle violations
- Missing licence headers
- Logging to console instead of file
- Missing Jira issue
- Commented out UriEndpoint annotation (+category attribute)
- Full build needed, to generate endpoint-dsl, component-dsl, ...

File: components/camel-oaipmh/src/main/java/org/apache/camel/oaipmh/component/OAIPMHProducer.java [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+    }
+
+    private void overrideHarvesterConfigs(Message msg, Harvester harvester) {
+        String header = msg.getHeader("CamelOaimphUrl", String.class);
+        if (header != null) {
+            harvester.setBaseURI(URI.create(header));
+        }
+
+        header = msg.getHeader("CamelOaimphEndpointUrl", String.class);
+        if (header != null) {
```
Comment (omarsmak) [inline components/camel-oaipmh/src/main/java/org/apache/camel/oaipmh/component/OAIPMHProducer.java:74]: I think these kind of validations, could be simplified into a function that takes lambda as argument, example:
```
private <T> void checkAndSetConfigs(final Message message, final String key, final Consumer<T> fn, final Class<T> type) {
        final T header = message.getHeader(key, type);
        if (!ObjectHelper.isEmpty(header)) {
            fn.accept(header);
        }
    }
``` 
to call it:
```
checkAndSetConfigs(msg, "CamelOaimphVerb", harvester::setVerb, String.class);
```
For the URI, you may need to add the a wrapper from URI to String somewhere that creates `URI.create(header)`

File: components/camel-oaipmh/src/main/java/org/apache/camel/oaipmh/utils/OAIPMHHttpClient.java [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+
+public class OAIPMHHttpClient {
+
+    private static final Logger LOG = LoggerFactory.getLogger(OAIPMHHttpClient.class);
+
+    private boolean ignoreSSLWarnings;
+
+    public String doRequest(URI baseURI, String verb, String set, String from, String until, String metadataPrefix, String token, String identifier) throws IOException, URISyntaxException, Exception {
+        CloseableHttpClient httpclient = getCloseableHttpClient();
+        try {
```
Comment (omarsmak) [inline components/camel-oaipmh/src/main/java/org/apache/camel/oaipmh/utils/OAIPMHHttpClient.java:48]: You can replace it with:
```suggestion
       try (CloseableHttpClient httpclient = getCloseableHttpClient()) {
```

File: components/camel-oaipmh/src/main/java/org/apache/camel/oaipmh/utils/OAIPMHHttpClient.java [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+            URIBuilder builder = new URIBuilder();
+            builder.setScheme(baseURI.getScheme())
+                    .setHost(baseURI.getHost())
+                    .setPort(baseURI.getPort())
+                    .setPath(baseURI.getPath())
+                    .addParameter("verb", verb);
+
+            if (baseURI.getQuery() != null && !baseURI.getQuery().isEmpty()) {
+                for (String param : baseURI.getQuery().split("&")) {
+                    builder.addParameter(param.split("=")[0], param.split("=")[1]);
```
Comment (omarsmak) [inline components/camel-oaipmh/src/main/java/org/apache/camel/oaipmh/utils/OAIPMHHttpClient.java:59]: I am just wondering, do we need to check that we really have two params split with `=`?

File: components/camel-oaipmh/src/main/java/org/apache/camel/oaipmh/utils/OAIPMHHttpClient.java [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+                        return EntityUtils.toString(entity, Charset.forName("UTF-8"));
+                    } else {
+                        throw new ClientProtocolException("Unexpected response status: " + status);
+                    }
+                }
+
+            };
+            String responseBody = httpclient.execute(httpget, responseHandler);
+
+//            String uri = requestLine.getUri();
```
Comment (omarsmak) [inline components/camel-oaipmh/src/main/java/org/apache/camel/oaipmh/utils/OAIPMHHttpClient.java:110]: unnecessary comments here

File: components/camel-oaipmh/src/main/java/org/apache/camel/oaipmh/utils/OAIPMHHttpClient.java [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+    protected CloseableHttpClient getCloseableHttpClient() throws Exception {
+        if (isIgnoreSSLWarnings()) {
+            try {
+                SSLContextBuilder builder = new SSLContextBuilder();
+                builder.loadTrustMaterial(null, new TrustSelfSignedStrategy());
+                SSLConnectionSocketFactory sslsf = new SSLConnectionSocketFactory(
+                        builder.build());
+                return HttpClients.custom().setSSLSocketFactory(
+                        sslsf).build();
+            } catch (Exception ex) {
```
Comment (omarsmak) [inline components/camel-oaipmh/src/main/java/org/apache/camel/oaipmh/utils/OAIPMHHttpClient.java:134]: If you are catching the exception and then throw it again, makes no sense. However, I don't feel comfortable of throwing generic `Exception` here, please be specific on which exception you want to throw here

File: components/camel-oaipmh/src/main/java/org/apache/camel/oaipmh/utils/OAIPMHHttpClient.java [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+            return responseBody;
+        } finally {
+            httpclient.close();
+        }
+    }
+
+    protected CloseableHttpClient getCloseableHttpClient() throws Exception {
+        if (isIgnoreSSLWarnings()) {
+            try {
+                SSLContextBuilder builder = new SSLContextBuilder();
```
Comment (omarsmak) [inline components/camel-oaipmh/src/main/java/org/apache/camel/oaipmh/utils/OAIPMHHttpClient.java:128]: This looks deprecated.

**Comment by juanksega** [comment]
Hi @bedlaj @omarsmak @DenisIstomin 

Thanks so much for your feedback. I think I fixed all your suggestions. Please, let me know if you have other suggestions.

**Comment by omarsmak** [comment]
@oscerd we shall merge this one?

**Comment by oscerd** [comment]
Yeah, let's merge and revisit and improving later

**Pull request merged**